### ISSUE TITLE
System.TypeInitializationException Unable to load DLL : sni.dll when attempting to get SQL data but DLL is present

### ISSUE THREAD

**Issue by dan-dragan** [aspnet/EntityFramework#4162]
**This is an issue in DNX and dotnet CLI (tracked by aspnet/dnx#3316 & dotnet/cli#881). We are leaving this issue open to track the high level scenario on the EF side and ensure everything works end-to-end for the RC2 release.**

<hr />

```
Stack trace:
Exception thrown: 'System.TypeInitializationException' in EntityFramework.Core.dll
Exception caught: The type initializer for 'System.Data.SqlClient.TdsParser' threw an exception.Exception caught: The type initializer for 'System.Data.SqlClient.SNILoadHandle' threw an exception.Exception caught: Unable to load DLL 'C:\Users\ddragan\.dnx\packages\runtime.win7-x64.System.Data.SqlClient.sni\4.0.0-beta-23616\runtimes\win7-x64\native\sni.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)Exception thrown: 'System.TypeInitializationException' in Platform.Authentication
Exception thrown: 'System.TypeInitializationException' in mscorlib.ni.dll
Microsoft.AspNet.Diagnostics.ExceptionHandlerMiddleware:Error: An unhandled exception has occurred: The type initializer for 'System.Data.SqlClient.TdsParser' threw an exception.
System.TypeInitializationException: The type initializer for 'System.Data.SqlClient.TdsParser' threw an exception. ---> System.TypeInitializationException: The type initializer for 'System.Data.SqlClient.SNILoadHandle' threw an exception. ---> System.DllNotFoundException: Unable to load DLL 'C:\Users\ddragan\.dnx\packages\runtime.win7-x64.System.Data.SqlClient.sni\4.0.0-beta-23616\runtimes\win7-x64\native\sni.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
   at System.Runtime.Loader.AssemblyLoadContext.InternalLoadUnmanagedDllFromPath(String unmanagedDllPath)
   at Microsoft.Dnx.Runtime.Loader.PackageAssemblyLoader.LoadUnmanagedLibrary(String name)
   at Microsoft.Dnx.Host.LoaderContainer.LoadUnmanagedLibrary(String name)
   at Microsoft.Dnx.Runtime.Loader.LoadContext.LoadUnmanagedDll(String unmanagedDllName)
   at System.Runtime.Loader.AssemblyLoadContext.ResolveUnmanagedDll(String unmanagedDllName, IntPtr gchManagedAssemblyLoadContext)
   at System.Data.SqlClient.SNINativeMethodWrapper.SNIInitialize(IntPtr pmo)
   at System.Data.SqlClient.SNILoadHandle..ctor()
   at System.Data.SqlClient.SNILoadHandle..cctor()
   --- End of inner exception stack trace ---
   at System.Data.SqlClient.TdsParser..cctor()
   --- End of inner exception stack trace ---
   at ExternalDataAccessor.GetFirm(String firmName)
```
runtime info :
```
D:\Work\repo\Web>dnvm list

Active Version           Runtime Architecture OperatingSystem Alias
------ -------           ------- ------------ --------------- -----
       1.0.0-beta5       clr     x64          win
       1.0.0-beta5       clr     x86          win
       1.0.0-beta5       coreclr x64          win
       1.0.0-beta5       coreclr x86          win
       1.0.0-beta6       clr     x64          win
       1.0.0-beta6       clr     x86          win
       1.0.0-beta6       coreclr x64          win
       1.0.0-beta6       coreclr x86          win
       1.0.0-beta8       clr     x64          win
       1.0.0-beta8       clr     x86          win
       1.0.0-beta8       coreclr x64          win
       1.0.0-beta8       coreclr x86          win
       1.0.0-rc1-final   clr     x86          win
       1.0.0-rc1-final   coreclr x64          win
       1.0.0-rc1-final   coreclr x86          win
       1.0.0-rc1-update1 clr     x64          win
       1.0.0-rc1-update1 clr     x86          win             default
       1.0.0-rc1-update1 coreclr x64          win
       1.0.0-rc1-update1 coreclr x86          win
       1.0.0-rc2-16258   clr     x86          win
```

Using VS2015 debugger: 
![dnx runtime](https://cloud.githubusercontent.com/assets/10272624/11968000/2938bb62-a8cd-11e5-8948-18a7cb33f8e4.png)

project.json content:
```json
{
  "webroot": "wwwroot",
  "version": "1.0.0-beta1-*",

  "content": "EnvironmentSettings.json",

  "dependencies": {
   
    "Platform.AspNet.Authentication.Diagnostics": "1.0.0-beta1-*",
    "Platform.AspNet.Authentication.DPerms": "1.0.0-beta1-*",
    "Platform.AspNet.Discovery": "1.0.0-beta1-*",
    "Platform.AspNet.Monitoring": "1.0.0-beta1-*"
  },

  "commands": {
    "kestrel": "Microsoft.AspNet.Hosting --server Microsoft.AspNet.Server.Kestrel --server.urls http://*:5004"
  },

  "frameworks": {
    "dnxcore50": {
      "dependencies": {
        "Microsoft.AspNet.Mvc": "6.0.0-rc1-final",
        "Microsoft.AspNet.Mvc.Core": "6.0.0-rc1-final",
        "Microsoft.AspNet.Mvc.ViewFeatures": "6.0.0-rc1-final",
        "Microsoft.AspNet.Mvc.TagHelpers": "6.0.0-rc1-final",
        "Microsoft.AspNet.IISPlatformHandler": "1.0.0-rc1-final",
        "Microsoft.AspNet.Server.WebListener": "1.0.0-rc1-final",
        "Microsoft.Extensions.DependencyInjection.Abstractions": "1.0.0-rc1-final",
        "Platform.Authentication": "1.0.0-rc1-final",
        "Microsoft.Extensions.OptionsModel": "1.0.0-rc1-final",
        "EntityFramework.Core": "7.0.0-rc1-final",
        "Microsoft.AspNet.Authentication": "1.0.0-rc1-final",
        "Microsoft.AspNet.Server.Kestrel": "1.0.0-rc1-final",
        "Microsoft.AspNet.StaticFiles": "1.0.0-rc1-final",
        "Microsoft.AspNet.FileProviders.Physical": "1.0.0-rc1-final",
        "Microsoft.Extensions.Configuration.Binder": "1.0.0-rc1-final",
        "Microsoft.AspNet.Hosting.Abstractions": "1.0.0-rc1-final",
        
        "Microsoft.AspNet.Identity": "3.0.0-rc1-final",
        "Microsoft.AspNet.Identity.EntityFramework": "3.0.0-rc1-final",
        "Microsoft.Extensions.Logging.Console": "1.0.0-rc1-final",
        "Microsoft.Extensions.Logging.Debug": "1.0.0-rc1-final",
        "Microsoft.VisualStudio.Web.BrowserLink.Loader": "14.0.0-rc1-final",
        "Microsoft.AspNet.Diagnostics": "1.0.0-rc1-final",
        "System.Linq": "4.0.1-beta-*"
      }
    }
  },

  "publishExclude": [
    "node_modules",
    "bower_components",
    "**.xproj",
    "**.user",
    "**.vspscc"
  ],
  "exclude": [
    "wwwroot",
    "node_modules",
    "bower_components"
  ]
}
```
File exists:
![fileexists](https://cloud.githubusercontent.com/assets/10272624/11968025/7a9e0a16-a8cd-11e5-8ca4-1781a5c1679e.png)

Thanks for looking into this:

**Comment by dan-dragan** [comment]
:-(

**Comment by natemcmaster** [comment]
Duplicate of https://github.com/aspnet/EntityFramework/issues/3787 ?

**Comment by dan-dragan** [comment]
Don't think so. Firstly I have VS2012 installed and so its runtime is there so the workaround is in place . Second, unlike what the other issues say, in my case, the dll is not missing but actually there.

Thanks for looking into this, I am blocked on a live project.

<div>-------- Original message --------</div><div>From: Nate McMaster <notifications@github.com> </div><div>Date:12/30/2015  5:41 PM  (GMT-06:00) </div><div>To: aspnet/EntityFramework <EntityFramework@noreply.github.com> </div><div>Cc: dan-dragan <dan@1dragan.com> </div><div>Subject: Re: [EntityFramework] System.TypeInitializationException Unable to
  load DLL : sni.dll when attempting to get SQL data but DLL is present (#4162) </div><div>
</div>Duplicate of #3787 ?

**Comment by natemcmaster** [comment]
@dan-dragan The error most likely comes from not having the C++ redistributable . It could be that the x64 version is missing. It appears you are using .NET Core x86 on the command line and x64 in Visual Studio. Can you please ensure x86 and x64 are installed? https://www.microsoft.com/en-us/download/details.aspx?id=30679

**Comment by dan-dragan** [comment]
As I said, I had them both but did install again from the linked location
![msvcrt2012x32installed](https://cloud.githubusercontent.com/assets/10272624/12069186/a2d38eee-afe8-11e5-817b-695d05eeb8f5.jpg)
![msvcrt2012x64installed](https://cloud.githubusercontent.com/assets/10272624/12069187/a2d446ae-afe8-11e5-91e3-19f7ea480868.jpg)

same result. Is it a matter of access rights perhaps? As I said, the file is in place.

**Comment by juswes** [comment]
Good day, I'm still aslo having the exact same issue. The sni.dll file is indeed pressend, but somehow is not exesable when I run my web app with IIS. 
But I would like to add that this issue only occurs if I run the web app with IIS. 
If I run the web app on the production server with the “web” command manually (after stopping it in IIS), then I can access the web app and get database connections without any issues. 

I’m running on coreclr-x64-rc1-update1 with IIS 7.5 on Windows Server 2008 R2 with C++ redistributable installed.

**Comment by natemcmaster** [comment]
@juswes your issue may be the same as https://github.com/dotnet/corefx/issues/4503 (resolved with https://github.com/aspnet/Hosting/issues/481)

Also, please note that code and issues for System.Data.SqlClient live in https://github.com/dotnet/corefx. There have been many bug fixes for SqlClient since RC1. These fixes are available in the RC2 nightly builds.

There is no actionable item for the Entity Framework.

**Comment by rowanmiller** [comment]
From the stack trace it looks like this is actually an EF scenario and not just raw SQL Client

> Exception thrown: 'System.TypeInitializationException' in EntityFramework.Core.dll

Even if the underlying issue is in SQL Client, we should still keep this active to track making sure the EF scenario gets fixed.

**Comment by natemcmaster** [comment]
More info: I can only reproduce "Exception thrown: 'System.TypeInitializationException' in EntityFramework.Core.dll" in Win7 by running without C++ redistributable. This issue continues appear for some users even with vcredist installed. It could be a DNX issue, IIS and permissions, PATH problems, or something else. I have been unable to reproduce the error with C++ redistributable installed.

> I am blocked on a live project.

@dan-dragan have you tried using full .NET instead? In VS, this is ".NET Framework" instead of ".NET Core". Using dnvm, this can be changed by executing `dnvm use 1.0.0-rc1-update1 -r clr`

**Comment by dan-dragan** [comment]
Thanks again for looking into this matter. No, I have not tried using full .NET, since I need multiple OS deployment capabilities. I will try it and let you know  - for the purpose of troubleshooting ....

**Comment by natemcmaster** [comment]
> I need multiple OS deployment capabilities

@dan-dragan heads up: SqlClient on .NET Core for Linux has an [issue with MARS](https://github.com/dotnet/corefx/issues/4487) and [with package restore](https://github.com/aspnet/EntityFramework/pull/3689#issuecomment-158201904). RC2 should fix these issues.

**Comment by dan-dragan** [comment]
Ack, thanks for the heads up.

**Comment by dan-dragan** [comment]
I thought I've already added this, but turns out I did not : dependency walker on the sni.dll in questions. All is *not* good :-( 
![sni dll_dependencies](https://cloud.githubusercontent.com/assets/10272624/12099654/71661804-b2ef-11e5-92ae-3298abe92e1d.jpg)

**Comment by natemcmaster** [comment]
@dan-dragan the "missing" dependencies you are seeing in dependency walker are in `%USERPROFILE%/.dnx/runtimes/dnx-coreclr-win-x64.1.0.0-rc1-update1/bin/` and are loaded by DNX.

 cc @victorhurdugaci @Tratcher do we have known downlevel issues in hosting/dnx regarding loading these runtime dependencies?

**Comment by dan-dragan** [comment]
@natemcmaster well, that sinks one promising theory/explanation...

**Comment by rowanmiller** [comment]
Hey,

Could someone who is seeing this issue confirm that they see it with just SQL Client (and no EF). Here is some sample code you could run to use SQL Client against LocalDb. You shouldn't need to install any packages as SQL Client is brought in by EF.

```c#
using (var conn = new SqlConnection(@"Server=(localdb)\mssqllocaldb;Database=master;Trusted_Connection=True;"))
{
    var cmd = conn.CreateCommand();
    cmd.CommandText = "SELECT COUNT(*) FROM sys.objects";
    conn.Open();
    var count = cmd.ExecuteScalar();
    conn.Close();
}
```

~Rowan

**Comment by JohnBicknell** [comment]
@rowanmiller I've pretty much exactly the same issue & setup as @juswes - coreclr-x64-rc1-update1 with IIS 7.5 on Windows Server 2008 R2 with C++ redistributable installed (x64 & x86).

I'm just using SQL Client directly no EF. 

SQL calls run fine if I start the app on the server with dnx web on localhost:5000. 
When I publish and try using IIS, SNI.dll issue occurs. 

Before installing the C++ redistributable, SNI.dll issue happened with both dnx web and IIS.
After installing the redistributable, dnx web was fixed but not IIS.

Tried with  1.0.0-rc2-16357   coreclr x64     with no success.

**Comment by rowanmiller** [comment]
Opened https://github.com/dotnet/corefx/issues/5252 to move this on to SQL Client's plate as there is nothing EF related here. Leaving this issue open though as this is a critical scenario for us and we want to track it to make sure it is fixed for RC2.

**Comment by dan-dragan** [comment]
So after reading through all the threads that this one was forked into my original issue seems to be a runtime PATH resolution problem.The location of the <correct runtime  for your case> is not added to the PATH , which prevents loading of sni.dll because some dependencies of it - in my case API-MS-Win-Core-StringAnsi-L1-1-0.dll - cannot be located.

Workarounds (thanks to @jamesgibbs100) are:

1) Copy the dependency dll(s) (that are matching your project settings (version/platform) as in coreclr-win-x64.1.0.0-rc1-final) in your project root folder.

or
 
2)Add the runtime path (in my case : C:\Users\ddragan\.dnx\runtimes\dnx-coreclr-win-x64.1.0.0-rc1-final\bin ) to your path via command line tool before starting VS from the same command line instance (to preserve the setting).

I don't know what the final resolution might be, most threads that this one was referred to are closed now.

**Comment by rowanmiller** [comment]
BTW the issues currently tracking the root cause are aspnet/dnx#3316 and dotnet/cli#881

**Comment by natemcmaster** [comment]
According to this https://github.com/dotnet/cli/issues/881#issuecomment-175928663, the issue should be fixed in RC2. I am unable to verify due to https://github.com/dotnet/cli/issues/1043

**Comment by victorhurdugaci** [comment]
Can you please a little more details about how 1043 impacts this repo?

**Comment by victorhurdugaci** [comment]
Whoops, @natemcmaster, ignore my previous comment. I thought it was a different issue

**Comment by natemcmaster** [comment]
This still does not appear to be fixed. I can repro with dotnet cli. See https://github.com/dotnet/cli/issues/1210

**Comment by natemcmaster** [comment]
I verified it works with the C++ redist 2012 installed. This dependency will be removed with https://github.com/dotnet/corefx/issues/3760.

**Comment by rowanmiller** [comment]
To clarify, this issue (separate from the need to have the C++ redist installed) was specific to running DNX under IIS. @natemcmaster has verified that this is no longer an issue after moving to .NET CLI (which is replacing DNX for RC2).
